File: src/bot/volo_bot.py
import asyncio
import json
import logging
import os
from collections import defaultdict
from src.sinks.whisper_sink import WhisperSink
import discord
import yaml
from datetime import datetime, timedelta

DISCORD_CHANNEL_ID = int(os.getenv("DISCORD_CHANNEL_ID"))
TRANSCRIPTION_METHOD = os.getenv("TRANSCRIPTION_METHOD")
PLAYER_MAP_FILE_PATH = os.getenv("PLAYER_MAP_FILE_PATH")
GUILD_ID=int(os.getenv("GUILD_ID"))

logger = logging.getLogger(__name__)

class VoloBot(discord.Bot):
    def __init__(self, loop):
        intents = discord.Intents.default()
        intents.message_content = True  # If you're reading messages
        intents.members = True          # Needed for guild.members
        intents.guilds = True
        super().__init__(command_prefix="!", loop=loop,intents=intents,
                         activity=discord.CustomActivity(name='Transcribing Audio to Text'))
        self.guild_to_helper = {}
        self.guild_is_recording = {}
        self.guild_whisper_sinks = {}
        self.guild_whisper_message_tasks = {}
        self.player_map = {}
        self._is_ready = False
        if TRANSCRIPTION_METHOD == "openai":
            self.transcriber_type = "openai"
        else:
            self.transcriber_type = "local"
        if PLAYER_MAP_FILE_PATH:
            with open(PLAYER_MAP_FILE_PATH, "r", encoding="utf-8") as file:
                self.player_map = yaml.safe_load(file)


    async def on_ready(self):
        logger.info(f"Logged in as {self.user} to Discord.")
        self._is_ready = True
        self.loop.create_task(self.schedule_weekly_task())

    async def schedule_weekly_task(self):
        TARGET_HOUR = 23
        TARGET_MINUTE = 1
        TARGET_WEEKDAY=4
        while True:
            now = datetime.now()
            # # Calculate next Sunday at 14:30
            days_ahead = TARGET_WEEKDAY - now.weekday()
            if days_ahead <= 0:
                if TARGET_HOUR>now.hour:
                    days_ahead += 7  # Schedule for next Sunday

            next_run = now + timedelta(days=days_ahead)
            next_run = next_run.replace(hour=TARGET_HOUR, minute=TARGET_MINUTE, second=0, microsecond=0)

            wait_seconds = (next_run - now).total_seconds()
            print(f"Next scheduled task in {wait_seconds / 3600:.2f} hours.")
            await asyncio.sleep(wait_seconds)

            await self.do_scheduled_task()

    async def do_scheduled_task(self):
        try:
            self.guild_whisper_sinks[GUILD_ID].vc.play(discord.FFmpegPCMAudio(source="assets/event.mp3", **{
                                'options': '-vn',
                                'executable': os.path.join("ffmpeg", "ffmpeg.exe")
                                }), after=lambda e: print("Done playing"))
        except Exception as e:
            print(f"failed scheduled task: {e}")


    async def close_consumers(self):
        await self.consumer_manager.close()
    def _close_and_clean_sink_for_guild(self, guild_id: int):
        whisper_sink: WhisperSink | None = self.guild_whisper_sinks.get(
            guild_id, None)

        if whisper_sink:
            logger.debug(f"Stopping whisper sink, requested by {guild_id}.")
            whisper_sink.stop_voice_thread()
            del self.guild_whisper_sinks[guild_id]
            whisper_sink.close()

    
    def start_recording(self):
        """
        Start recording audio from the voice channel. Create a whisper sink
        and start sending transcripts to the queue.

        Since this is a critical function, this is where we should handle
        subscription checks and limits.
        """
        try:
            self.start_whisper_sink()
            self.guild_is_recording[GUILD_ID] = True
        except Exception as e:
            logger.error(f"Error starting whisper sink: {e}")

    def start_whisper_sink(self):
        guild_voice_sink = self.guild_whisper_sinks.get(GUILD_ID, None)
        if guild_voice_sink:
            logger.debug(
                f"Sink is already active for guild {GUILD_ID}.")
            return

        async def on_stop_record_callback(sink: WhisperSink):
            logger.debug(
                f"{GUILD_ID} -> on_stop_record_callback")
            self._close_and_clean_sink_for_guild(GUILD_ID)

        transcript_queue = asyncio.Queue()

        whisper_sink = WhisperSink(
            transcript_queue,
            self.loop,
            data_length=50000,
            max_speakers=7,
            transcriber_type=self.transcriber_type,
            player_map=self.player_map,
            bot=self
        )
        self.guild_to_helper[GUILD_ID].vc.start_recording(
            whisper_sink, on_stop_record_callback)
        def on_thread_exception(e):
            logger.warning(
                f"Whisper sink thread exception for guild {GUILD_ID}. Retry in 5 seconds...\n{e}")
            self._close_and_clean_sink_for_guild(GUILD_ID)

            # retry in 5 seconds
            self.loop.call_later(5, self.start_recording)
        whisper_sink.start_voice_thread(on_exception=on_thread_exception)

        self.guild_whisper_sinks[GUILD_ID] = whisper_sink

    def stop_recording(self, ctx: discord.context.ApplicationContext):
        vc = ctx.guild.voice_client
        if vc:
            self.guild_is_recording[ctx.guild_id] = False
            vc.stop_recording()
        guild_id = ctx.guild_id
        whisper_message_task = self.guild_whisper_message_tasks.get(
            guild_id, None)
        if whisper_message_task:
            logger.debug("Cancelling whisper message task.")
            whisper_message_task.cancel()
            del self.guild_whisper_message_tasks[guild_id]
    def cleanup_sink(self, ctx: discord.context.ApplicationContext):
        guild_id = ctx.guild_id
        self._close_and_clean_sink_for_guild(guild_id)

    async def get_transcription(self, ctx: discord.context.ApplicationContext):
        # Get the transcription queue
        if not (self.guild_whisper_sinks.get(ctx.guild_id)):
            return
        whisper_sink = self.guild_whisper_sinks[ctx.guild_id]
        transcriptions = []
        if whisper_sink is None:
            return
    
        transcriptions_queue = whisper_sink.transcription_output_queue
        while not transcriptions_queue.empty():
            transcriptions.append(await transcriptions_queue.get())
        return transcriptions

    async def update_player_map(self, ctx: discord.context.ApplicationContext):
        player_map = {}
        for member in ctx.guild.members:
            player_map[member.id] = {
                "player": member.name,
                "character": member.display_name
            }
        logger.info(f"{str(player_map)}")
        self.player_map.update(player_map)
        if PLAYER_MAP_FILE_PATH:
            with open(PLAYER_MAP_FILE_PATH, "w", encoding="utf-8") as file:
                yaml.dump(self.player_map, file, default_flow_style=False, allow_unicode=True)

    async def stop_and_cleanup(self):
        try:
            for sink in self.guild_whisper_sinks.values():
                sink.close()
                sink.stop_voice_thread()
                logger.debug(
                    f"Stopped whisper sink for guild {sink.vc.channel.guild.id} in cleanup.")
            self.guild_whisper_sinks.clear()
        except Exception as e:
            logger.error(f"Error stopping whisper sinks: {e}")
        finally:
            logger.info("Cleanup completed.")
    

File: src/bot/volo_bot.ts
import { Client, GatewayIntentBits, ActivityType, VoiceChannel, VoiceState, Guild, Collection } from 'discord.js';
import { VoiceConnectionStatus } from '@discordjs/voice';
import { BotHelper } from './helper';
import { WhisperSink } from '../sinks/whisper_sink';
import * as yaml from 'js-yaml';
import * as fs from 'fs';
import * as path from 'path';
import { ChatInputCommandInteraction, EmbedBuilder } from 'discord.js';

const TRANSCRIPTION_METHOD = process.env.TRANSCRIPTION_METHOD || "local";
const PLAYER_MAP_FILE_PATH = process.env.PLAYER_MAP_FILE_PATH;
const GUILD_ID = process.env.GUILD_ID!;

export class VoloBot extends Client {
  guildToHelper: Map<string, BotHelper> = new Map();
  guildIsRecording: Map<string, boolean> = new Map();
  guildWhisperSinks: Map<string, WhisperSink> = new Map();
  guildWhisperMessageTasks: Map<string, any> = new Map();
  playerMap: { [key: string]: { player: string; character: string } } = {};
  _isReady: boolean = false;
  transcriberType: string;

  constructor() {
    super({
      intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMessages,
        GatewayIntentBits.MessageContent,
        GatewayIntentBits.GuildMembers,
        GatewayIntentBits.GuildVoiceStates
      ],
      presence: {
        activities: [{
          name: 'Transcribing Audio to Text',
          type: ActivityType.Custom
        }]
      }
    });

    this.transcriberType = TRANSCRIPTION_METHOD === "openai" ? "openai" : "local";
    
    if (PLAYER_MAP_FILE_PATH && fs.existsSync(PLAYER_MAP_FILE_PATH)) {
      try {
        const fileContents = fs.readFileSync(PLAYER_MAP_FILE_PATH, 'utf8');
        this.playerMap = yaml.load(fileContents) as { [key: string]: { player: string; character: string } };
      } catch (e) {
        console.error(`Error loading player map: ${e}`);
      }
    }
  }

  async onReady(): Promise<void> {
    console.log(`Logged in as ${this.user?.tag} to Discord.`);
    this._isReady = true;
    this.scheduleWeeklyTask();
  }

  async scheduleWeeklyTask(): Promise<void> {
    const TARGET_HOUR = 23;
    const TARGET_MINUTE = 1;
    const TARGET_WEEKDAY = 4; // Friday

    while (true) {
      const now = new Date();
      let daysAhead = TARGET_WEEKDAY - now.getDay();
      
      if (daysAhead <= 0) {
        if (TARGET_HOUR > now.getHours()) {
          daysAhead += 7;
        }
      }

      const nextRun = new Date(now);
      nextRun.setDate(now.getDate() + daysAhead);
      nextRun.setHours(TARGET_HOUR, TARGET_MINUTE, 0, 0);

      const waitSeconds = (nextRun.getTime() - now.getTime()) / 1000;
      console.log(`Next scheduled task in ${waitSeconds / 3600:.2f} hours.`);
      
      await new Promise(resolve => setTimeout(resolve, waitSeconds * 1000));
      await this.doScheduledTask();
    }
  }

  async doScheduledTask(): Promise<void> {
    try {
      const sink = this.guildWhisperSinks.get(GUILD_ID);
      if (sink && sink.vc) {
        // Play audio file
        const audioPath = path.join(process.cwd(), "assets", "event.mp3");
        if (fs.existsSync(audioPath)) {
          // Audio playback would be handled by the voice connection
          // This is a simplified version - actual implementation depends on discord.js voice
          console.log("Playing scheduled audio");
        }
      }
    } catch (e: any) {
      console.error(`Failed scheduled task: ${e}`);
    }
  }

  _closeAndCleanSinkForGuild(guildId: string): void {
    const whisperSink = this.guildWhisperSinks.get(guildId);

    if (whisperSink) {
      console.log(`Stopping whisper sink, requested by ${guildId}.`);
      whisperSink.stopVoiceThread();
      this.guildWhisperSinks.delete(guildId);
      whisperSink.close();
    }
  }

  startRecording(): void {
    try {
      this.startWhisperSink();
      this.guildIsRecording.set(GUILD_ID, true);
    } catch (e: any) {
      console.error(`Error starting whisper sink: ${e}`);
    }
  }

  startWhisperSink(): void {
    const guildVoiceSink = this.guildWhisperSinks.get(GUILD_ID);
    if (guildVoiceSink) {
      console.log(`Sink is already active for guild ${GUILD_ID}.`);
      return;
    }

    const helper = this.guildToHelper.get(GUILD_ID);
    if (!helper || !helper.vc) {
      console.error("No voice connection available for guild");
      return;
    }

    const transcriptQueue = new Array(); // Simplified queue for now

    const whisperSink = new WhisperSink(
      transcriptQueue,
      this,
      {
        dataLength: 50000,
        maxSpeakers: 7,
        transcriberType: this.transcriberType,
        playerMap: this.playerMap
      }
    );

    whisperSink.vc = helper.vc;

    // Set up audio receiver to capture audio from users
    const receiver = helper.vc.receiver;
    
    helper.vc.on('stateChange', (oldState, newState) => {
      if (newState.status === VoiceConnectionStatus.Ready) {
        // Start listening to all users in the voice channel
        const guild = this.guilds.cache.get(GUILD_ID);
        if (guild) {
          const voiceChannel = helper.vc?.joinConfig.channelId 
            ? guild.channels.cache.get(helper.vc.joinConfig.channelId) as any
            : null;
          
          if (voiceChannel) {
            voiceChannel.members.forEach((member: any) => {
              if (!member.user.bot) {
                this.startListeningToUser(receiver, member.id, whisperSink);
              }
            });
          }
        }
      }
    });

    const onThreadException = (e: any) => {
      console.warn(`Whisper sink thread exception for guild ${GUILD_ID}. Retry in 5 seconds...\n${e}`);
      this._closeAndCleanSinkForGuild(GUILD_ID);
      setTimeout(() => this.startRecording(), 5000);
    };

    whisperSink.startVoiceThread(onThreadException);
    this.guildWhisperSinks.set(GUILD_ID, whisperSink);
  }

  private startListeningToUser(receiver: VoiceReceiver, userId: string, sink: WhisperSink): void {
    const audioStream = receiver.subscribe(userId, {
      end: {
        behavior: EndBehaviorType.AfterSilence,
        duration: 100
      }
    });

    audioStream.on('data', (chunk: Buffer) => {
      sink.write(chunk, userId);
    });
  }

  stopRecording(ctx: ChatInputCommandInteraction): void {
    const vc = ctx.guild?.voiceStates.cache.get(this.user!.id)?.channel;
    if (vc) {
      this.guildIsRecording.set(ctx.guildId!, false);
      // Stop recording would be handled by the voice connection
    }

    const guildId = ctx.guildId!;
    const whisperMessageTask = this.guildWhisperMessageTasks.get(guildId);
    if (whisperMessageTask) {
      console.log("Cancelling whisper message task.");
      clearTimeout(whisperMessageTask);
      this.guildWhisperMessageTasks.delete(guildId);
    }
  }

  cleanupSink(ctx: ChatInputCommandInteraction): void {
    const guildId = ctx.guildId!;
    this._closeAndCleanSinkForGuild(guildId);
  }

  async getTranscription(ctx: ChatInputCommandInteraction): Promise<any[]> {
    const whisperSink = this.guildWhisperSinks.get(ctx.guildId!);
    if (!whisperSink) {
      return [];
    }

    const transcriptions: any[] = [];
    const transcriptionsQueue = whisperSink.transcriptionOutputQueue;
    
    while (transcriptionsQueue.length > 0) {
      transcriptions.push(transcriptionsQueue.shift()!);
    }
    
    return transcriptions;
  }

  async updatePlayerMap(ctx: ChatInputCommandInteraction): Promise<void> {
    const playerMap: { [key: string]: { player: string; character: string } } = {};
    
    const guild = await this.guilds.fetch(ctx.guildId!);
    const members = await guild.members.fetch();
    
    members.forEach(member => {
      playerMap[member.id] = {
        player: member.user.username,
        character: member.displayName
      };
    });

    console.log(JSON.stringify(playerMap));
    this.playerMap = { ...this.playerMap, ...playerMap };

    if (PLAYER_MAP_FILE_PATH) {
      try {
        fs.writeFileSync(PLAYER_MAP_FILE_PATH, yaml.dump(this.playerMap, { defaultFlowStyle: false, allowUnicode: true }), 'utf8');
      } catch (e) {
        console.error(`Error writing player map: ${e}`);
        throw e;
      }
    }
  }

  async stopAndCleanup(): Promise<void> {
    try {
      for (const sink of this.guildWhisperSinks.values()) {
        sink.close();
        sink.stopVoiceThread();
        console.log(`Stopped whisper sink in cleanup.`);
      }
      this.guildWhisperSinks.clear();
    } catch (e: any) {
      console.error(`Error stopping whisper sinks: ${e}`);
    } finally {
      console.log("Cleanup completed.");
    }
  }
}

